subsys/net/ip/tcp.c

```diff
@@ -569,7 +569,20 @@ static int tcp_conn_unref(struct tcp *conn)
 
 	tcp_send_queue_flush(conn);
 
-	k_work_cancel_delayable(&conn->send_data_timer);
+	/* Cancel all possible delayed work and prevent any execution of newly scheduled work
+	 * in one of the work item handlers
+	 * Essential because the work items are destructed at the bottom of this method.
+	 * A current ongoing execution might access the connection and schedule new work
+	 * Solution:
+	 * While holding the lock, cancel all delayable work.
+	 * Because every delayable work execution takes the same lock and releases the lock,
+	 * we're either here, or currently executing one of the workers.
+	 * Then, after cancelling the workers within the lock, either those workers have finished
+	 * or have been cancelled and will not execute anymore
+	 */
+	k_mutex_lock(&conn->lock, K_FOREVER);
```

Comment:
I wonder whether it'd make sense to use `k_work_cancel_delayable_sync()` to cancel works instead, which should wait until the work has finished. If we take the mutex here, and in the meantime some other work is scheduled and blocks at the mutex, it'd still execute after the mutex is released (and possibly reschedule some work).

At the same time, work handlers could check the connection state, and simply return early if the connection is already in `TCP_CLOSED` state, to prevent rescheduling works.

Reply:
Nope, tried that k_work_cancel_delayable_sync(), but without the mutex being taken. I think I saw a deadlock, because the k_mutex_lock(&tcp_lock, K_FOREVER); is taken, and if the work-item is then executed on this context, and that mutex is taken by one of the work handlers, you'd have a deadlock (unless it's a counting mutex).
At least what I witnessed it that the mqtt client, using a tcp connection, lost the connection (as expected by scenario) and that never recovered anymore.

Reply:
`At the same time, work handlers could check the connection state, and simply return early if the connection is already in TCP_CLOSED state, to prevent rescheduling works.`

Thought about that too but didn't know that to use for checks. But those aren't needed in my solution, I simply don't care if they execute or not, I just make sure that they never are invoked anymore after I destructed the connection.

Reply:
> Nope, tried that k_work_cancel_delayable_sync(), but without the mutex being taken. I think I saw a deadlock

Ah, I think the issue here (which I did not consider) is that the connection can be unref'ed from the work handler itself, so if we block until work is done it may indeed end up in a deadlock.